### ISSUE TITLE
dep file parsing doesn't look at the first line, but the first line can have files on it

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#2046]
https://github.com/ziglang/zig/blob/fec4555476e38d2eeb1dfb02572404b243acd0b2/src/cache_hash.cpp#L431-L432

But here's an example file:

```
test.o: test.c b/foo.h
```

In this example there are 2 files on the first line. The dep file parsing probably needs to gain a real tokenizer, and support any number of items on any line. For example it also doesn't look for multiple files on the same line if there are double quotes.

I also haven't managed to find documentation for this file format. It would be nice to find a reference. Supposedly it's "NMake or Jom" format.

Finally, audit the places that call `cache_add_dep_file` and make sure they don't redundantly add the main source file, since it's contained in the dep file as well.

**Comment by surajbarkale** [comment]
Technically dep file must be a valid Makefile. In practice, compilers support a few different forms. Ninja build has a comprehensive parser for the format: https://github.com/ninja-build/ninja/blob/master/src/depfile_parser.in.cc

The supported constructs end up being:
1. Simple
```Make
    out: in1 in2 in3
```
2. Escaped newlines
```Make
    out: in1 \
         in2 \
         in3
```
3. Separate declarations
```Make
    out: in1
    out: in2
    out: in3
```
4. Ignore phony rules generated by the `gcc -MP` option
```Make
    out: in1 in2 in3
    in1:
    in2:
    in3:
```

**Comment by mikdusan** [comment]
> I also haven't managed to find documentation for this file format

file `clang/lib/Frontend/DependencyFile.cpp` function `PrintFilename()` is probably the best we're going to get. what I can see right now is if someone is doing trickery with quotes inside an include file name, clang's emission doesn't handle it too well so they're really not all too formalized anyways.

Here is output from both styles, using **2** source files `foo.c` is simple, and `odd.....name.c` is including a bunch of specials. One surprise is that the target name uses make/compatible format even if `-MV` is passed to clang. So I'm thinking it would probably just be easier to implement a tokenizer for make/compatible because NMake/Jom format would require a tokenizer that understood make/compatible anyways.

#### NMake/Jom format
#### `clang -MV -MM -MF - -MG foo.c 'odd{ball} #$^! name.c'`
```
foo.o: foo.c foo_01.h foo_02.h foo_03.h foo_04.h foo_05.h foo_06.h \
  foo_07.h foo_08.h foo_09.h foo_10.h
odd{ball}\ \#$$^!\ name.o: "odd{ball} #$^! name.c" "odd{01}.h" "odd$.h" "odd^.h" \
  "odd#.h" "odd!.h"
```

#### make/compatible format - notice esca
#### `clang -MM -MF - -MG foo.c 'odd{ball} #$^! name.c'`
```
foo.o: foo.c foo_01.h foo_02.h foo_03.h foo_04.h foo_05.h foo_06.h \
  foo_07.h foo_08.h foo_09.h foo_10.h
odd{ball}\ \#$$^!\ name.o: odd{ball}\ \#$$^!\ name.c odd{01}.h odd$$.h odd^.h \
  odd\#.h odd!.h
```

**Comment by andrewrk** [comment]
On Windows, the `\` is used in file paths, so it can't be relied on for escaping. The make/compatible format is ambiguous on Windows.

**Comment by mikdusan** [comment]
>Finally, audit the places that call `cache_add_dep_file` and make sure they don't redundantly add the main source file, since it's contained in the dep file as well.

is this to mean that in the following `.d`, `source.c` should **not** be passed to `cache_add_file()` because the build system already knows the dep?
```make
zig-cache/tmp/h6XndQRnlXKV-source.o: source.c foo.h bar.h
```

**Comment by andrewrk** [comment]
Yep I think you have the correct understanding. Stated another way, it means that in this case `cache_add_dep_file` should be assumed to add `source.c`, `foo.h`, and `bar.h`, and so an additional `cache_add_file` for `source.c` would be redundant.

**Comment by mikdusan** [comment]
> On Windows, the `\` is used in file paths, so it can't be relied on for escaping. The make/compatible format is ambiguous on Windows.
> 
> Let's have a chat with the clang developers and make sure that this isn't a bug, both the ambiguous case for make/compatible format, and the case where the target name does not respect `-MV`.

hold off a bit. i've got what I think is a really good implementation that i'm fuzz-testing using default make/compatible .d format.

as it relates to make/compatible format, It took me a bit to get these nuggets:

- '\' is not a general escape
- '\' **must** be followed by one of { <space>, '#', '\' }
- if it is not followed by one of these 3 chars, it is an error
- so windows paths end up in the make file as `\\temp\\foo.h`
- and my tokenizer translates it to raw bytes `\temp\foo.h`
- related: '$' is not a general escape
- '$' **must** be followed by exactly one '$' otherwise an error
- similarly, tokenizer translates it `filename$$.h` → `filename$.h`

**Comment by andrewrk** [comment]
Without `-MV`, windows paths don't get escaped. 
![bad-dep-file](https://user-images.githubusercontent.com/106511/55349242-30597d00-5487-11e9-8564-cc131d2e6306.png)
This format is ambiguous.

**Comment by mikdusan** [comment]
> This format is ambiguous.

So that answers my question. clang was re-writing `\` → `/` on macOS. I guess it's rewriting as per test-obj.d file.

But -MV is still ambiguous too because (at least on macOS) it does this:

```
foo$$.o: "foo$.c" "foo$.h"
```

target does not obey -MV syntax. it's prereqs do. I understand that it's less likely a target will need quotation. 

Right now it sounds the like the best way to go is switch prereq tokens (RHS of colon) to -MV format. And leave make/compatible tokenization for target tokens (LHS of colon). Just be aware we still have that ambiguity with target.

**Comment by andrewrk** [comment]
I think it's definitely worth filing a bug report with clang. I can help with that.

**Comment by mikdusan** [comment]
`find` and `xargs` have a need to share a file list and they use `-print0` and `-0` to pipe nullz filenames. Piping aside (that's not a terrible idea either but feature creep), a new `-M0` or `-MZ` option for clang would suit: 3 targets represented fairly easily:

```
target,null
prereq,null
null
target,null
null
target,null
prereq,null
prereq,null
```

**Comment by andrewrk** [comment]
> But -MV is still ambiguous too because (at least on macOS) it does this:

```
foo$$.o: "foo$.c" "foo$.h"
```

Is that actually ambiguous? Isn't the `$` escaped with another `$`? We need to find out the behavior on Windows.

**Comment by andrewrk** [comment]
Upstream bug report: https://bugs.llvm.org/show_bug.cgi?id=41379

**Comment by mikdusan** [comment]
> > But -MV is still ambiguous too because (at least on macOS) it does this:
> 
> ```
> foo$$.o: "foo$.c" "foo$.h"
> ```
> Is that actually ambiguous? Isn't the `$` escaped with another `$`? We need to find out the behavior on Windows.

Yes, `foo$$.o` is not ambiguous, my intent was only to show that simultaneously, an emitted `.d` file has two kinds of escaping going on.

**Comment by andrewrk** [comment]
Goodness. Did you see this comment?

https://github.com/llvm/llvm-project/blob/ce2d45e7ba4d86a1f7b694deefddda4b60ff6fd7/clang/lib/Frontend/DependencyFile.cpp#L352-L399

What a mess.

**Comment by mikdusan** [comment]
For posterity: on Windows here is what clang does with the following **target** pathnames; note that drive letter `:` is also never escaped. I'll have to get creative because it clashes with target-end delimiter.

literal pathname|.d target|number of literal spaces
---|---|:-:
`C:\Users\mike\foo.o`|`C:\Users\mike\foo.o:`
`C:\Users\mike\foo .o`|`C:\Users\mike\foo\ .o:`|1
`C:\Users\mike\foo#.o`|`C:\Users\mike\foo\#.o:`
`C:\Users\mike\foo$.o`|`C:\Users\mike\foo$$.o:`
`C:\Users\mike\ foo.o`|`C:\Users\mike\\\ foo.o:`|1
`C:\Users\mike\#foo.o`|`C:\Users\mike\\#foo.o:`
`C:\Users\mike\$foo.o`|`C:\Users\mike\$$foo.o:`
`C:\Users\mike\     foo.o`|`C:\Users\mike\\\ \ \ \ \ foo.o:`|5

**Comment by andrewrk** [comment]
Now that @mikdusan's new parser is merged, Zig looks at the first line of dep file parsing but this issue remains, because clang outputs ambiguous .d files. So this issue is open to follow up with the [upstream bug report](https://bugs.llvm.org/show_bug.cgi?id=41379) and get the issue fixed in clang.

One more item to solve, once the first line files is solved, is making sure that zig does not redundantly add the target file to the cache. This would be harmless, but a small waste of time.

**Comment by andrewrk** [comment]
Upstream bug report has moved to GitHub: https://github.com/llvm/llvm-project/issues/40724

This does not appear to be a high priority for the LLVM team so will likely require us to send a patch in order to get addressed.
